rts/idris_rts.c

```diff
@@ -411,102 +435,193 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
+        exit(EXIT_FAILURE);
+    }
+    // Then grab the UTF8 bytes out of the Idris string. We don't have
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    i_int v = strtol(temp, &end, 10);
+    Closure* answer;
+    // If we used the whole string, return the int, otherwise return 0. This is
+    // the meaning of the primitive in the evaluator.
+    if (end == temp + str.size)
+        answer = MKINT(v);
     else
-        return MKINT(0);
+        answer = MKINT(0);
+    free(temp);
+    return answer;
 }
 
 VAL idris_castFloatStr(VM* vm, VAL i) {
-    Closure* cl = allocate(sizeof(Closure) + sizeof(char)*32, 0);
-    SETTY(cl, STRING);
-    cl -> info.str = (char*)cl + sizeof(Closure);
-    snprintf(cl -> info.str, 32, "%.16g", GETFLOAT(i));
-    return cl;
+    // Use a maximum of 31 digits to print floats here, plus 1 char for null
+    const unsigned int MAX_CHARS_FOR_DOUBLE = 32;
+    char temp[MAX_CHARS_FOR_DOUBLE];
+    memset(temp, '\0', MAX_CHARS_FOR_DOUBLE);
+
+    snprintf(temp, MAX_CHARS_FOR_DOUBLE, "%.16g", GETFLOAT(i));
+
+    // avoid invalid UTF-8: all ASCII digits, plus period, plus, and
+    // minus, are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < MAX_CHARS_FOR_DOUBLE; j++) {
+        if (temp[j] != '\0' &&
+            temp[j] != '.' &&
+            temp[j] != '-' &&
+            temp[j] != '+' &&
+            !(isdigit(temp[j]))
+           ) {
+            fprintf(stderr, "Fatal error: invalid output from sprintf was %s!", temp);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+
+    return MKSTR(vm, (utf8_byte *)temp);
 }
 
 VAL idris_castStrFloat(VM* vm, VAL i) {
-    return MKFLOAT(vm, strtod(GETSTR(i), NULL));
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough for the
+    // data and the null at the end
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
+        exit(EXIT_FAILURE);
+    }
+    // Then grab the UTF8 bytes out of the Idris string. We don't have
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    char *end;
+    double v = strtod(temp, &end);
+    Closure* answer;
+    // If we used the whole string, return the double, otherwise
+    // return 0.0. This is the meaning of the primitive in the
+    // evaluator.
```

Comment:
It is, however, the current semantic of the primitive. Changing that is a separate issue.